File: mobile/lib/core/services/auth_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_auth/firebase_auth.dart';

import '../../models/user_model.dart';
import '../constants/firebase_constants.dart';

// ─── Custom exception ─────────────────────────────────────────────────────────

class AuthException implements Exception {
  final String message;
  const AuthException(this.message);

  @override
  String toString() => message;
}

// ─── Auth Service ─────────────────────────────────────────────────────────────

class AuthService {
  AuthService({
    FirebaseAuth? auth,
    FirebaseFirestore? firestore,
  })  : _auth = auth ?? FirebaseAuth.instance,
        _firestore = firestore ?? FirebaseFirestore.instance;

  final FirebaseAuth _auth;
  final FirebaseFirestore _firestore;

  /// Emits whenever the auth state changes (login / logout).
  Stream<User?> get authStateChanges => _auth.authStateChanges();

  /// Synchronous reference to the currently signed-in Firebase user.
  User? get currentUser => _auth.currentUser;

  // ── Sign In ────────────────────────────────────────────────────────────────

  Future<UserModel> signIn({
    required String email,
    required String password,
  }) async {
    try {
      final credential = await _auth.signInWithEmailAndPassword(
        email: email.trim(),
        password: password,
      );

      final uid = credential.user!.uid;
      final doc = await _firestore
          .collection(FirebaseConstants.usersCollection)
          .doc(uid)
          .get();

      if (!doc.exists) {
        throw const AuthException(
          'Account data not found. Please contact support.',
        );
      }

      return UserModel.fromFirestore(doc);
    } on FirebaseAuthException catch (e) {
      throw AuthException(_mapFirebaseError(e.code));
    } catch (_) {
      throw const AuthException(
          'Sign-in failed. Please check your connection and try again.');
    }
  }

  // ── Register ───────────────────────────────────────────────────────────────

  /// Creates a Firebase Auth user and writes all Firestore documents
  /// (users + role-specific profile) in a single batch.
  Future<UserModel> register({
    required String email,
    required String password,
    required String fullName,
    required UserRole role,
    required String phoneNumber,
    required String country,
    Map<String, dynamic>? roleData,
  }) async {
    try {
      final credential = await _auth.createUserWithEmailAndPassword(
        email: email.trim(),
        password: password,
      );

      final uid = credential.user!.uid;
      final now = DateTime.now();

      final user = UserModel(
        id: uid,
        email: email.trim(),
        fullName: fullName.trim(),
        role: role,
        phoneNumber: phoneNumber.trim(),
        country: country,
        verificationStatus: VerificationStatus.pending,
        createdAt: now,
        updatedAt: now,
      );

      final batch = _firestore.batch();

      // ── users/{uid} ────────────────────────────────────────────────────────
      final userRef =
          _firestore.collection(FirebaseConstants.usersCollection).doc(uid);
      batch.set(userRef, user.toFirestore());

      // ── Role-specific profile ──────────────────────────────────────────────
      switch (role) {
        case UserRole.ofw:
          final ofwRef = _firestore
              .collection(FirebaseConstants.ofwProfilesCollection)
              .doc();
          final ofw = OFWProfile(
            id: ofwRef.id,
            userId: uid,
            passportNumber: roleData?['passport_number'] as String?,
            destinationCountry: roleData?['destination_country'] as String?,
            jobTitle: roleData?['job_title'] as String?,
            experienceYears: roleData?['experience_years'] as int?,
            contractStatus: ContractStatus.none,
            createdAt: now,
          );
          batch.set(ofwRef, ofw.toFirestore());

        case UserRole.agency:
          final agencyRef =
              _firestore.collection(FirebaseConstants.agenciesCollection).doc();
          final agency = AgencyModel(
            id: agencyRef.id,
            userId: uid,
            agencyName: roleData?['agency_name'] as String? ?? '',
            licenseNumber: roleData?['license_number'] as String? ?? '',
            country: roleData?['country'] as String? ?? country,
            address: roleData?['address'] as String? ?? '',
            contactEmail: roleData?['contact_email'] as String? ?? email.trim(),
            contactPhone:
                roleData?['contact_phone'] as String? ?? phoneNumber.trim(),
            verificationStatus: 'pending',
            createdAt: now,
          );
          batch.set(agencyRef, agency.toFirestore());

        case UserRole.verifier:
          final verifierRef = _firestore
              .collection(FirebaseConstants.verifierProfilesCollection)
              .doc();
          final orgTypeStr =
              roleData?['organization_type'] as String? ?? 'government';
          final verifier = VerifierProfile(
            id: verifierRef.id,
            userId: uid,
            organizationName: roleData?['organization_name'] as String? ?? '',
            organizationType: OrganizationType.values.firstWhere(
              (t) => t.name == orgTypeStr,
              orElse: () => OrganizationType.government,
            ),
            roleTitle: roleData?['role_title'] as String? ?? '',
            verificationLevel: 1,
            createdAt: now,
          );
          batch.set(verifierRef, verifier.toFirestore());

        default:
          break;
      }

      await batch.commit();
      return user;
    } on FirebaseAuthException catch (e) {
      throw AuthException(_mapFirebaseError(e.code));
    } catch (_) {
      throw const AuthException(
          'Registration failed. Please check your connection and try again.');
    }
  }

  // ── Load current user ──────────────────────────────────────────────────────

  /// Fetches the Firestore [UserModel] for the currently signed-in Firebase
  /// user, or returns null if nobody is logged in.
  Future<UserModel?> loadCurrentUser() async {
    final fbUser = _auth.currentUser;
    if (fbUser == null) return null;
    final doc = await _firestore
        .collection(FirebaseConstants.usersCollection)
        .doc(fbUser.uid)
        .get();
    if (!doc.exists) return null;
    return UserModel.fromFirestore(doc);
  }

  // ── Sign Out ───────────────────────────────────────────────────────────────

  Future<void> signOut() => _auth.signOut();

  // ── Helpers ────────────────────────────────────────────────────────────────

  String _mapFirebaseError(String code) {
    switch (code) {
      case 'user-not-found':
        return 'No account found with this email.';
      case 'wrong-password':
      case 'invalid-credential':
        return 'Incorrect email or password.';
      case 'email-already-in-use':
        return 'An account with this email already exists.';
      case 'weak-password':
        return 'Password is too weak. Use at least 6 characters.';
      case 'invalid-email':
        return 'Please enter a valid email address.';
      case 'user-disabled':
        return 'This account has been disabled. Contact support.';
      case 'too-many-requests':
        return 'Too many failed attempts. Please try again later.';
      case 'network-request-failed':
        return 'No internet connection. Please check your network.';
      default:
        return 'An unexpected error occurred. Please try again.';
    }
  }
}

File: web/src/core/services/auth.service.ts
import {
  createUserWithEmailAndPassword,
  signInWithEmailAndPassword,
  signOut as firebaseSignOut,
  updateProfile,
} from "firebase/auth";
import { collection, doc, getDoc, writeBatch } from "firebase/firestore";
import { auth, db } from "../config/firebase";
import {
  USERS_COLLECTION,
  OFW_PROFILES_COLLECTION,
  AGENCIES_COLLECTION,
  VERIFIER_PROFILES_COLLECTION,
} from "../constants/firebase";
import type { UserModel, UserRole, OrganizationType } from "@/types";
import {
  userModelFromFirestore,
  userModelToFirestoreData,
  ofwProfileToFirestoreData,
  agencyModelToFirestoreData,
  verifierProfileToFirestoreData,
  USER_ROLE,
  VERIFICATION_STATUS,
  CONTRACT_STATUS,
  ORGANIZATION_TYPE,
} from "@/types";

// ─── Custom exception ─────────────────────────────────────────────────────────

export class AuthException extends Error {
  constructor(message: string) {
    super(message);
    this.name = "AuthException";
  }
}

// ─── Error mapping ───────────────────────────────────────────────────────────

function mapFirebaseError(code: string): string {
  // Firebase JS SDK often prefixes error codes, e.g. "auth/wrong-password".
  // Normalise to the final segment so we can switch on stable codes.
  const normalized = code.includes("/")
    ? (code.split("/").pop() ?? code)
    : code;

  switch (normalized) {
    case "user-not-found":
      return "No account found with this email.";
    case "wrong-password":
    case "invalid-login-credentials":
    case "invalid-credential":
      return "Incorrect email or password.";
    case "email-already-in-use":
      return "An account with this email already exists.";
    case "weak-password":
      return "Password is too weak. Use at least 6 characters.";
    case "invalid-email":
      return "Please enter a valid email address.";
    case "user-disabled":
      return "This account has been disabled. Contact support.";
    case "too-many-requests":
      return "Too many failed attempts. Please try again later.";
    case "network-request-failed":
      return "No internet connection. Please check your network.";
    case "permission-denied":
      return "You don't have permission to access your account data. Please contact support.";
    default:
      return "An unexpected error occurred. Please try again.";
  }
}

// ─── Sign In ─────────────────────────────────────────────────────────────────

/**
 * Signs in with email/password, then fetches the Firestore user document.
 * Returns UserModel or throws AuthException.
 */
export async function signIn(
  email: string,
  password: string,
): Promise<UserModel> {
  try {
    const credential = await signInWithEmailAndPassword(
      auth,
      email.trim(),
      password,
    );
    const uid = credential.user.uid;
    const userRef = doc(db, USERS_COLLECTION, uid);
    const userSnap = await getDoc(userRef);
    if (!userSnap.exists()) {
      throw new AuthException(
        "Account data not found. Please contact support.",
      );
    }
    return userModelFromFirestore(userSnap);
  } catch (err) {
    if (err instanceof AuthException) throw err;
    const code =
      err && typeof err === "object" && "code" in err
        ? String((err as { code: string }).code)
        : "";
    throw new AuthException(mapFirebaseError(code || "unknown"));
  }
}

// ─── Register ───────────────────────────────────────────────────────────────

export type RegisterRoleData = {
  // OFW
  passport_number?: string;
  destination_country?: string;
  job_title?: string;
  experience_years?: number;
  // Agency
  agency_name?: string;
  license_number?: string;
  country?: string;
  address?: string;
  contact_email?: string;
  contact_phone?: string;
  // Verifier
  organization_name?: string;
  organization_type?: string;
  role_title?: string;
};

/**
 * Creates Firebase Auth user and writes users + role-specific profile in a batch.
 * Returns UserModel or throws AuthException.
 */
export async function register(
  email: string,
  password: string,
  fullName: string,
  role: UserRole,
  phoneNumber: string,
  country: string,
  roleData?: RegisterRoleData,
): Promise<UserModel> {
  try {
    const credential = await createUserWithEmailAndPassword(
      auth,
      email.trim(),
      password,
    );
    const uid = credential.user.uid;
    const displayName = fullName.trim();
    if (displayName) {
      await updateProfile(credential.user, { displayName });
    }

    const now = new Date();
    const user: UserModel = {
      id: uid,
      email: email.trim(),
      fullName: displayName || email.trim(),
      role,
      phoneNumber: phoneNumber.trim(),
      country,
      profilePhotoUrl: null,
      verificationStatus: VERIFICATION_STATUS.PENDING,
      createdAt: now,
      updatedAt: now,
    };

    const batch = writeBatch(db);

    // users/{uid}
    const userRef = doc(db, USERS_COLLECTION, uid);
    batch.set(userRef, userModelToFirestoreData(user));

    // Role-specific profile
    const rd = roleData ?? {};
    switch (role) {
      case USER_ROLE.OFW: {
        const ofwRef = doc(collection(db, OFW_PROFILES_COLLECTION));
        batch.set(
          ofwRef,
          ofwProfileToFirestoreData({
            id: ofwRef.id,
            userId: uid,
            passportNumber: rd.passport_number ?? null,
            destinationCountry: rd.destination_country ?? null,
            jobTitle: rd.job_title ?? null,
            experienceYears: rd.experience_years ?? null,
            agencyId: null,
            contractStatus: CONTRACT_STATUS.NONE,
            createdAt: now,
          }),
        );
        break;
      }
      case USER_ROLE.AGENCY: {
        const agencyRef = doc(collection(db, AGENCIES_COLLECTION));
        batch.set(
          agencyRef,
          agencyModelToFirestoreData({
            id: agencyRef.id,
            userId: uid,
            agencyName: rd.agency_name ?? "",
            licenseNumber: rd.license_number ?? "",
            country: rd.country ?? country,
            address: rd.address ?? "",
            contactEmail: rd.contact_email ?? email.trim(),
            contactPhone: rd.contact_phone ?? phoneNumber.trim(),
            verificationStatus: "pending",
            createdAt: now,
          }),
        );
        break;
      }
      case USER_ROLE.VERIFIER: {
        const verifierRef = doc(collection(db, VERIFIER_PROFILES_COLLECTION));
        const orgType =
          (rd.organization_type as OrganizationType) ??
          ORGANIZATION_TYPE.GOVERNMENT;
        batch.set(
          verifierRef,
          verifierProfileToFirestoreData({
            id: verifierRef.id,
            userId: uid,
            organizationName: rd.organization_name ?? "",
            organizationType: orgType,
            roleTitle: rd.role_title ?? "",
            verificationLevel: 1,
            createdAt: now,
          }),
        );
        break;
      }
      default:
        break;
    }

    await batch.commit();
    return user;
  } catch (err) {
    if (err instanceof AuthException) throw err;
    const code =
      err && typeof err === "object" && "code" in err
        ? String((err as { code: string }).code)
        : "";
    throw new AuthException(mapFirebaseError(code || "unknown"));
  }
}

// ─── Sign Out ───────────────────────────────────────────────────────────────

export async function signOut(): Promise<void> {
  await firebaseSignOut(auth);
}
